<a href="https://colab.research.google.com/github/Ramya1410/Bangalore-House-Price-Prediction/blob/main/Bangalore_House_Price_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('Bengaluru_House_Data.csv')

In [ ]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [ ]:
df.shape

(13320, 9)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [ ]:
from tables import Column
for col in df.columns:
  print(df[col].value_counts())
  print('*'*25)

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64
*************************
Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
15-Aug               1
17-Jan               1
16-Nov               1
16-Jan               1
14-Jul               1
Name: availability, Length: 81, dtype: int64
*************************
Whitefield                        540
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Abshot Layout                       1
Name: location, Length: 1305, dtype: int64
*************************
2 BHK         519

checking for null values

In [ ]:
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

In [ ]:
0.25*13320

3330.0

the number of null values in 'society' field is more than 25% of the total rows. therefore, the column will be dropped.
removing cols that dont actually matter for the prediction

In [ ]:
df = df.drop(columns =['area_type','society','balcony','availability'])
df.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [ ]:
df.describe()

bath         price
count  13247.000000  13320.000000
mean       2.692610    112.565627
std        1.341458    148.971674
min        1.000000      8.000000
25%        2.000000     50.000000
50%        2.000000     72.000000
75%        3.000000    120.000000
max       40.000000   3600.000000

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13319 non-null  object 
 1   size        13304 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13247 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


In [ ]:
df['location']= df['location'].fillna('Sarjapur Road')

In [ ]:
df['size'].value_counts()

2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 BHK           2
11 Bedroom       2
10 BHK           2
14 BHK           1
13 BHK           1
12 Bedroom       1
27 BHK           1
43 Bedroom       1
16 BHK           1
19 BHK           1
18 Bedroom       1
Name: size, dtype: int64

In [ ]:
df['size']=df['size'].fillna('2 BHK')

In [ ]:
df['bath']=df['bath'].fillna(df['bath'].median())

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13320 non-null  object 
 1   size        13320 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13320 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


hence all null values are cleared. we can move on.


in the size columns, we can see that some are bhk and som eare bedroom. so let us make it numerical.

In [ ]:
df['BHK'] = df['size'].str.split().str.get(0).astype(int)

In [ ]:
df[df.BHK >20]

location        size total_sqft  bath  price  BHK
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

these two are ouliers and need to be removed.

In [ ]:
df['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

the values in 'total_sqft' should be either int or float values. but there are some ranges included. so , we need to fix that.

we are going to use the hiphen between the two values and replace the range with the mean of the 2 values.

In [ ]:
def convertRange(x):
  temp = x.split('-')
  if len(temp) == 2:
    return (float(temp[0]) + float(temp[1]))/2
  try:
    return float(x)
  except: 
    return None

In [ ]:
df['total_sqft'] = df['total_sqft'].apply(convertRange)

In [ ]:
df.head()

location       size  total_sqft  bath   price  BHK
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

In [ ]:
df.rename(columns = {'price': 'price in lakhs'}, inplace = True)

In [ ]:
df.head()

location       size  total_sqft  bath  price in lakhs  BHK
0  Electronic City Phase II      2 BHK      1056.0   2.0           39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0          120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0           62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0           95.00    3
4                  Kothanur      2 BHK      1200.0   2.0           51.00    2

Let us create a new column called 'price per square feet'. it helps in removing outliers.

In [ ]:
df['price_per_sqft'] = df['price in lakhs'] * 100000 / df['total_sqft']

In [ ]:
df['price_per_sqft']

0         3699.810606
1         4615.384615
2         4305.555556
3         6245.890861
4         4250.000000
             ...     
13315     6689.834926
13316    11111.111111
13317     5258.545136
13318    10407.336319
13319     3090.909091
Name: price_per_sqft, Length: 13320, dtype: float64

In [ ]:
df.describe()

total_sqft          bath  price in lakhs           BHK  \
count  13274.000000  13320.000000    13320.000000  13320.000000   
mean    1559.626694      2.688814      112.565627      2.802778   
std     1238.405258      1.338754      148.971674      1.294496   
min        1.000000      1.000000        8.000000      1.000000   
25%     1100.000000      2.000000       50.000000      2.000000   
50%     1276.000000      2.000000       72.000000      3.000000   
75%     1680.000000      3.000000      120.000000      3.000000   
max    52272.000000     40.000000     3600.000000     43.000000   

       price_per_sqft  
count    1.327400e+04  
mean     7.907501e+03  
std      1.064296e+05  
min      2.678298e+02  
25%      4.266865e+03  
50%      5.434306e+03  
75%      7.311746e+03  
max      1.200000e+07

In [ ]:
df['location'].value_counts()

Whitefield                        540
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Uvce Layout                         1
Abshot Layout                       1
Name: location, Length: 1306, dtype: int64

we can see that there are 1306 different locations. so, we cannot one hot encode it. so , let us clean this by putting any location that is less than 10 to a location called 'other'.

to remove the spaces before and after the actual location, we are using strip function.

In [ ]:
df['location']= df['location'].apply(lambda x: x.strip())
location_count = df['location'].value_counts()

In [ ]:
location_count

Whitefield                            541
Sarjapur  Road                        399
Electronic City                       304
Kanakpura Road                        273
Thanisandra                           237
                                     ... 
1Channasandra                           1
Hosahalli                               1
Vijayabank bank layout                  1
near Ramanashree California resort      1
Abshot Layout                           1
Name: location, Length: 1295, dtype: int64

In [ ]:
location_count_lessthan_10 = location_count[location_count<=10]
location_count_lessthan_10

BTM 1st Stage                         10
Nagadevanahalli                       10
Basapura                              10
Sector 1 HSR Layout                   10
Dairy Circle                          10
                                      ..
1Channasandra                          1
Hosahalli                              1
Vijayabank bank layout                 1
near Ramanashree California resort     1
Abshot Layout                          1
Name: location, Length: 1054, dtype: int64

we can see that there are 1054 locations that have count <= 10

In [ ]:
df['location'] = df['location'].apply(lambda x: 'other' if x in location_count_lessthan_10 else x)

In [ ]:
df['location'].value_counts()

other                 2886
Whitefield             541
Sarjapur  Road         399
Electronic City        304
Kanakpura Road         273
                      ... 
Nehru Nagar             11
Banjara Layout          11
LB Shastri Nagar        11
Pattandur Agrahara      11
Narayanapura            11
Name: location, Length: 242, dtype: int64

therefore we can see that the length reduced and hence now, one hot encoder can be applied.

# outlier detection and removal

In [ ]:
df.describe()

total_sqft          bath  price in lakhs           BHK  \
count  13274.000000  13320.000000    13320.000000  13320.000000   
mean    1559.626694      2.688814      112.565627      2.802778   
std     1238.405258      1.338754      148.971674      1.294496   
min        1.000000      1.000000        8.000000      1.000000   
25%     1100.000000      2.000000       50.000000      2.000000   
50%     1276.000000      2.000000       72.000000      3.000000   
75%     1680.000000      3.000000      120.000000      3.000000   
max    52272.000000     40.000000     3600.000000     43.000000   

       price_per_sqft  
count    1.327400e+04  
mean     7.907501e+03  
std      1.064296e+05  
min      2.678298e+02  
25%      4.266865e+03  
50%      5.434306e+03  
75%      7.311746e+03  
max      1.200000e+07

here we can see that in total_sqft, the min value is one. its logically and physically not possible to have a place of 1sqft. therefore, it is definitely an outlier and needs to be rectified.

In [ ]:
(df['total_sqft']/df['BHK']).describe()

count    13274.000000
mean       575.074878
std        388.205175
min          0.250000
25%        473.333333
50%        552.500000
75%        625.000000
max      26136.000000
dtype: float64

In [ ]:
df = df[((df['total_sqft']/df['BHK']) >= 300)]

In [ ]:
df.describe()

total_sqft          bath  price in lakhs           BHK  \
count  12530.000000  12530.000000    12530.000000  12530.000000   
mean    1594.564544      2.559537      111.382401      2.650838   
std     1261.271296      1.077938      152.077329      0.976678   
min      300.000000      1.000000        8.440000      1.000000   
25%     1116.000000      2.000000       49.000000      2.000000   
50%     1300.000000      2.000000       70.000000      3.000000   
75%     1700.000000      3.000000      115.000000      3.000000   
max    52272.000000     16.000000     3600.000000     16.000000   

       price_per_sqft  
count    12530.000000  
mean      6303.979357  
std       4162.237981  
min        267.829813  
25%       4210.526316  
50%       5294.117647  
75%       6916.666667  
max     176470.588235

now we can observe that the min value is 300 

In [ ]:
df.shape

(12530, 7)

In [ ]:
df.price_per_sqft.describe()

count     12530.000000
mean       6303.979357
std        4162.237981
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_per_sqft, dtype: float64

here, if we observe the 'max' value, we can see that it is an outlier cuz its too expensive for a sqft and not possible.

In [ ]:
def remove_outliers_sqft(df):
  df_output = pd.DataFrame()
  for key, subdf in df.groupby('location'):
    m = np.mean(subdf.price_per_sqft)
    st = np.std(subdf.price_per_sqft)
    gen_df = subdf[(subdf.price_per_sqft > (m-st)) & (subdf.price_per_sqft < (m+st))]
    df_output = pd.concat([df_output,gen_df], ignore_index=True)
  return df_output
df = remove_outliers_sqft(df)
df.describe()

total_sqft          bath  price in lakhs           BHK  \
count  10301.000000  10301.000000    10301.000000  10301.000000   
mean    1508.440608      2.471702       91.286372      2.574896   
std      880.694214      0.979449       86.342786      0.897649   
min      300.000000      1.000000       10.000000      1.000000   
25%     1110.000000      2.000000       49.000000      2.000000   
50%     1286.000000      2.000000       67.000000      2.000000   
75%     1650.000000      3.000000      100.000000      3.000000   
max    30400.000000     16.000000     2200.000000     16.000000   

       price_per_sqft  
count    10301.000000  
mean      5659.062876  
std       2265.774749  
min       1250.000000  
25%       4244.897959  
50%       5175.600739  
75%       6428.571429  
max      24509.803922

here when groupby function is used, every location is taken as 'key' and data related to it beccomes the sub dataframe (subdf)

In [ ]:
def bhk_outliers_remove(df):
  exclude_indices = np.array([]) #this is an empty array we are creating to store the indices that are going to be removed based on the condition we will provide.
  for location, location_df in df.groupby('location'): #here location is 'key' and location_df is the sub df
    bhk_stats = {} #this is an empty dict that we're creating to store the stats of the different bhks.
    for bhk, bhk_df in df.groupby('BHK'): #here bhk is 'key' and bhk_df is the sub df
      bhk_stats[bhk] = {
          'mean' : np.mean(bhk_df.price_per_sqft),
          'std' : np.std(bhk_df.price_per_sqft),
          'count' : bhk_df.shape[0]
      }
    for bhk, bhk_df in df.groupby('BHK'):
      stats = bhk_stats.get(bhk-1)
      if stats and stats['count']>5:
        exclude_indices = np.append(exclude_indices, bhk_df[bhk_df.price_per_sqft < (stats['mean'])].index.values)
  return df.drop(exclude_indices, axis = 'index')


In [ ]:
df = bhk_outliers_remove(df)

In [ ]:
df.shape

(5385, 7)

In [ ]:
df

location       size  total_sqft  bath  price in lakhs  BHK  \
0      1st Block Jayanagar      4 BHK      2850.0   4.0           428.0    4   
1      1st Block Jayanagar      3 BHK      1630.0   3.0           194.0    3   
2      1st Block Jayanagar      3 BHK      1875.0   2.0           235.0    3   
3      1st Block Jayanagar      3 BHK      1200.0   2.0           130.0    3   
4      1st Block Jayanagar      2 BHK      1235.0   2.0           148.0    2   
...                    ...        ...         ...   ...             ...  ...   
10294                other      2 BHK      1095.0   2.0            57.0    2   
10295                other      2 BHK      1256.0   2.0            65.0    2   
10296                other      2 BHK      1353.0   2.0           110.0    2   
10297                other  1 Bedroom       812.0   1.0            26.0    1   
10300                other      4 BHK      3600.0   5.0           400.0    4   

       price_per_sqft  
0        15017.543860  
1        11901.840491  
2        12533.333333  
3        10833.333333  
4        11983.805668  
...               ...  
10294     5205.479452  
10295     5175.159236  
10296     8130.081301  
10297     3201.970443  
10300    11111.111111  

[5385 rows x 7 columns]

the columns size, price_per_sqft arent useful. so we will drop them. we created the price_per_sqft to help remove the outliers and clean the data.

In [ ]:
df.drop(columns = ['size', 'price_per_sqft'], inplace = True)

# cleaned data

In [ ]:
df.head()

location  total_sqft  bath  price in lakhs  BHK
0  1st Block Jayanagar      2850.0   4.0           428.0    4
1  1st Block Jayanagar      1630.0   3.0           194.0    3
2  1st Block Jayanagar      1875.0   2.0           235.0    3
3  1st Block Jayanagar      1200.0   2.0           130.0    3
4  1st Block Jayanagar      1235.0   2.0           148.0    2

* now our features are location, total_sqft, bath, and BHK. 
* our target variable is price in lakhs.

let us save our cleaned data separately.

In [ ]:
df.to_csv('cleaned_data.csv')

# label encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
encoder = LabelEncoder()
df['location'] = encoder.fit_transform(df['location'])

In [ ]:
X = df.drop(columns = ['price in lakhs']) #features
y = df['price in lakhs'] #target variable

In [ ]:
X = X.values
y = y.values

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2)

# Applying Linear Regression

In [ ]:
from sklearn import linear_model

In [ ]:
model = linear_model.LinearRegression()

In [ ]:
model.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
print(X_train.shape)
print(X_test.shape)

(4308, 4)
(1077, 4)


In [ ]:
r2_score = model.score(X_test,y_test)

In [ ]:
print(r2_score*100,'%')

78.20264715415306 %


# Applying Random Forest Regressor


In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
model=RandomForestRegressor(n_estimators=10, verbose=4)

In [ ]:
model.fit(X_train, y_train)

building tree 1 of 10
building tree 2 of 10
building tree 3 of 10
building tree 4 of 10
building tree 5 of 10
building tree 6 of 10
building tree 7 of 10
building tree 8 of 10
building tree 9 of 10
building tree 10 of 10


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.1s finished


RandomForestRegressor(n_estimators=10, verbose=4)

In [ ]:
y_pred = model.predict(X_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished


In [ ]:
r2_score = model.score(X_test,y_test)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.0s finished


In [ ]:
print(r2_score*100,'%')

78.9897158473001 %


# Applying decision Tree Regressor


In [ ]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
model = DecisionTreeRegressor()

In [ ]:
model.fit(X_train, y_train)

DecisionTreeRegressor()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
r2_score = model.score(X_test,y_test)

In [ ]:
print(r2_score*100,'%')

72.479144011518 %


# Applying KNeighors Regressor


In [ ]:
from sklearn.neighbors import KNeighborsRegressor

In [ ]:
model = KNeighborsRegressor(n_neighbors=10)

In [ ]:
model.fit(X_train, y_train)

KNeighborsRegressor(n_neighbors=10)

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
r2_score = model.score(X_test,y_test)

In [ ]:
print(r2_score*100,'%')

74.47147850721228 %


Hence, we can see that, Random Forest Regressor and Linear regressor gives maximum accuracy whereas K Neighbors Regressor and Decision Tree gives less accuracy.